src/mlpack/core/tree/spill_tree/spill_single_tree_traverser_impl.hpp

```diff
@@ -52,30 +52,27 @@ SpillSingleTreeTraverser<RuleType, Defeatist>::Traverse(
   if (referenceNode.IsLeaf())
   {
     for (size_t i = 0; i < referenceNode.NumPoints(); ++i)
-    {
-      size_t referenceIndex = referenceNode.Point(i);
-
-      // Don't run the basecase for the reference point if it is already present in
-      // the neighbor list of the querypoint.
-      if(!rule.InNeighborList(queryIndex, referenceIndex))
-        rule.BaseCase(queryIndex, referenceIndex);
-    }
+      rule.BaseCase(queryIndex, referenceNode.Point(i));
   }
   else
   {
     if (Defeatist && referenceNode.Overlap())
     {
       // If referenceNode is a overlapping node we do defeatist search.
       size_t bestChild = rule.GetBestChild(queryIndex, referenceNode);
-      Traverse(queryIndex, referenceNode.Child(bestChild));
 
-      // If All the neighbors have not been found recurse into the other child.
-      // This is the case when leaf size is less than the number of nearest
-      // neighbors to be found.
-      if(!rule.HasFoundAllNeighbors(queryIndex))
+      // If the bestchild node has less number of points than the minimum
+      // basecases to be run, recurse into other node as well.
+      if (referenceNode.Child(bestChild).NumPoints() <= rule.MinBaseCases())
+      {
+        Traverse(queryIndex, referenceNode.Child(bestChild));
```

Comment:
Realized my mistake. Here basecase will run more than one time for a point present in both child node.
I tried running the basecases without recursing but got wrong result.

Replacing
```
Traverse(queryIndex, referenceNode.Child(bestChild));
Traverse(queryIndex, referenceNode.Child(!bestChild));
```
with
```
for (size_t i = 0; i < referenceNode.NumPoints(); ++i)
      rule.BaseCase(queryIndex, referenceNode.Point(i));
```

I got distance matrix:
![Screenshot from 2020-06-21 18-13-43](https://user-images.githubusercontent.com/31977306/85224926-1920eb00-b3bd-11ea-98ee-535935377345.png)

Reply:
Ahh, so `NumPoints()` only refers to the points held in that node, not the points held in any child or descendant nodes.  Try using `i < referenceNode.NumDescendants()` and `referenceNode.Descendant(i)` instead---does that work?  I think it will at least get us closer.  If there are still duplicated points then I think I know what the issue is so I can help out in that case. :)

Reply:
I tried your suggestion. It only has duplicate points error. In my first commit I provided a way to handle duplicate points. Do you have any other method?

Reply:
Right, so the issue with the original commit you had was that it still calls `BaseCase()` multiple times, and to really fix this issue we should figure out how to avoid that.

Can you try replacing the implementation of `Descendant()` in `spill_tree_impl.hpp` with this?

```
template<typename MetricType,
         typename StatisticType,
         typename MatType,
         template<typename HyperplaneMetricType> class HyperplaneType,
         template<typename SplitMetricType, typename SplitMatType>
             class SplitType>
inline size_t SpillTree<MetricType, StatisticType, MatType, HyperplaneType,
    SplitType>::Descendant(const size_t index) const
{
  if (IsLeaf())
    return (*pointsIndex)[index];

  // Compute the number of points that are in both the left and right children.
  const size_t overlap = (left->NumDescendants() + right->NumDescendants()) - 
      count;

  const size_t num = left->NumDescendants();
  if (index < num)
    return left->Descendant(index);

  // Since points on the right may overlap, we have to start our count on the
  // right ignoring any overlapped points.
  if (right)
    return right->Descendant(index - num + overlap);

  // This should never happen.
  return (size_t() - 1);
}
```

I think that that will fix the duplicate point issue.  Basically, we want to expect that `referenceNode.Descendant(i)` does not return any duplicates, and I believe that once that is true, everything will work.

If that doesn't work, I'll check out the branch and start playing around and debugging. :+1:

Reply:
Still the same issue.
```
if (right)
    return right->Descendant(index - num + overlap);
```
I think this will only work when overlapped points are indexed before other points.

Reply:
Right, I *thought* that was the case but maybe not.  Let me dig in and try to figure out how to fix the spill tree implementation.  I'll post another comment shortly (hopefully! :)).

Reply:
Okay... I at least fixed the spill tree implementation:
https://gist.github.com/4dedacaaa36e33613e284d9614c0fec9
Try applying that patch to this branch and we can see if it solves the duplicate point issue.  At the very least, I think it fixes some strange behavior of the spill tree. :+1:

Reply:
Applied your patch. Last hunk was not applying so I added it manually. Still the issue of duplicate points though :(